File: x/src/main/java/com/sscl/baselibrary/files/FileUtil.java
package com.sscl.baselibrary.files;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Environment;
import android.util.Log;


import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;

/**
 * 文件工具类
 *
 * @author ALM
 */
public class FileUtil {

    /*--------------------------------静态常量--------------------------------*/

    /**
     * 在文件管理器中的程序文件夹名（默认值，未执行初始化函数就是这个值）
     */
    private static final String UNNAMED_APP_DIR = "UnnamedAppDir";

    /*--------------------------------成员变量--------------------------------*/

    /**
     * 初始化状态
     */
    private static boolean init = false;

    /**
     * 应用程序名(在文件管理器中的程序文件夹名)
     */
    private static String APP_NAME = UNNAMED_APP_DIR;

    /*--------------------------------公开静态方法--------------------------------*/

    /**
     * 初始化
     *
     * @param context 上下文
     */
    public static void init(@NonNull Context context) {
        init(context, true);
    }

    /**
     * 初始化
     *
     * @param context            上下文
     * @param withOutPackageName 是否需要包名
     */
    @SuppressWarnings("WeakerAccess")
    public static void init(@NonNull Context context, boolean withOutPackageName) {
        File filesDir = context.getFilesDir();
        String absolutePath = filesDir.getAbsolutePath();
        String[] split = absolutePath.split("/");
        String appName = split[split.length - 2];
        if (withOutPackageName) {
            String[] split1 = appName.split("\\.");
            appName = split1[split1.length - 1];
        }
        FileUtil.APP_NAME = appName;
        init = true;
    }

    /**
     * 获取SD卡文件目录
     *
     * @return 获取SD卡路径
     */
    @SuppressWarnings("WeakerAccess")
    @NonNull
    public static File getSdCardDir() {
        //获取挂载状态
        String state = Environment.getExternalStorageState();

        //如果是已挂载,说明了有内存卡
        if (Environment.MEDIA_MOUNTED.equals(state)) {
            return Environment.getExternalStorageDirectory();
        }
        throw new NullPointerException("No SD card was found!");
    }

    /**
     * 获取本项目文件目录
     *
     * @return 本项目文件目录
     */
    @Nullable
    public static File getAppDir() {

        if (!init) {
            throw new NullPointerException("FileUtil not init");
        }
        if (null == APP_NAME || "".equals(APP_NAME)) {
            throw new ExceptionInInitializerError("File name invalid");
        }
        File file = new File(getSdCardDir(), APP_NAME);
        boolean mkdirs;
        if (!file.exists()) {
            mkdirs = file.mkdirs();
            if (mkdirs) {
                return file;
            } else {
                return null;
            }
        }
        return file;
    }

    /**
     * 获取项目缓存文件目录
     *
     * @return 项目缓存文件目录
     */
    @SuppressWarnings("WeakerAccess")
    @Nullable
    public static File getCacheDir() {
        File file = new File(getAppDir(), "cache");
        boolean mkdirs;
        if (!file.exists()) {
            mkdirs = file.mkdirs();
            if (mkdirs) {
                return file;
            } else {
                return null;
            }
        }
        return file;
    }

    /**
     * 获取项目的异常日志目录
     *
     * @return 项目的异常日志目录
     */
    @Nullable
    public static File getCrashDir() {
        File file = new File(getAppDir(), "crash");
        boolean mkdirs;
        if (!file.exists()) {
            mkdirs = file.mkdirs();
            if (mkdirs) {
                return file;
            } else {
                return null;
            }
        }
        return file;
    }

    /**
     * 获取apk存储目录
     *
     * @return apk存储目录
     */
    @SuppressWarnings("unused")
    @Nullable
    public static File getApkDir() {
        File file = new File(getAppDir(), "apk");
        boolean mkdirs;
        if (!file.exists()) {
            mkdirs = file.mkdirs();
            if (mkdirs) {
                return file;
            } else {
                return null;
            }
        }
        return file;
    }

    /**
     * 获取图片目录
     *
     * @return 图片目录
     */
    @SuppressWarnings("unused")
    @Nullable
    public static File getImageDir() {
        File file = new File(getAppDir(), "images");
        boolean mkdirs;
        if (!file.exists()) {
            mkdirs = file.mkdirs();
            if (mkdirs) {
                return file;
            } else {
                return null;
            }
        }
        return file;
    }

    /**
     * 安装apk
     *
     * @param context 上下文
     * @param apkFile 文件
     */
    @SuppressWarnings("unused")
    public static void installApk(@NonNull Context context,@NonNull File apkFile) {
        Intent intent = new Intent();
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.setAction(Intent.ACTION_VIEW);
        String type = "application/vnd.android.package-archive";
        intent.setDataAndType(Uri.fromFile(apkFile), type);
        context.startActivity(intent);
    }

    /**
     * 获取项目数据目录
     *
     * @return 项目数据目录
     */
    @SuppressWarnings("unused")
    @Nullable
    public static File getDataDir() {
        File file = new File(getAppDir(), "data");
        boolean mkdirs;
        if (!file.exists()) {
            mkdirs = file.mkdirs();
            if (mkdirs) {
                return file;
            } else {
                return null;
            }
        }
        return file;
    }

    /**
     * 删除文件
     *
     * @param fileName 文件名
     */
    @SuppressWarnings("unused")
    public static boolean delFile(@NonNull String fileName) {
        File file = new File(getAppDir(), fileName);
        if (file.isFile()) {
            //noinspection ResultOfMethodCallIgnored
            file.delete();
            return true;
        }
        //noinspection ResultOfMethodCallIgnored
        file.exists();
        return false;
    }

    /**
     * 保存图片
     *
     * @param bm      位图图片
     * @param picName 文件名
     */
    @SuppressWarnings("unused")
    public static void saveBitmap(@NonNull Bitmap bm,@NonNull String picName) {
        Log.e("", "保存图片");
        try {

            File f = new File(getCacheDir(), picName + ".JPEG");
            if (f.exists()) {
                //noinspection ResultOfMethodCallIgnored
                f.delete();
            }

            FileOutputStream out = new FileOutputStream(f);
            bm.compress(Bitmap.CompressFormat.JPEG, 90, out);
            out.flush();
            out.close();
            Log.e("", "已经保存");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * 检查文件是否存在
     *
     * @param fileName 文件名
     * @return true表示文件存在
     */
    @SuppressWarnings("unused")
    public static boolean isFileExist(@NonNull String fileName) {
        File file = new File(getAppDir(), fileName);
        //noinspection ResultOfMethodCallIgnored
        file.isFile();
        return file.exists();
    }

    /**
     * 检查文件是否存在
     *
     * @param path 文件路径
     * @return true表示文件存在
     */
    @SuppressWarnings("unused")
    public static boolean fileIsExists(@NonNull String path) {
        try {
            File file = new File(path);
            if (!file.exists()) {
                return false;
            }
        } catch (Exception e) {

            return false;
        }
        return true;
    }
}

File: baselibrary/src/main/java/com/sscl/baselibrary/files/FileUtil.kt
package com.sscl.baselibrary.files

import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.util.Log
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

/**
 * 文件工具类
 *
 * @author ALM
 */
object FileUtil {

    /* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
     *
     * 属性声明
     *
     * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

    /* * * * * * * * * * * * * * * * * * * 私有常量 * * * * * * * * * * * * * * * * * * */

    /**
     * 在文件管理器中的程序文件夹名（默认值，未执行初始化函数就是这个值）
     */
    private const val UNNAMED_APP_DIR = "UnnamedAppDir"

    /* * * * * * * * * * * * * * * * * * * 私有可变属性 * * * * * * * * * * * * * * * * * * */

    /**
     * 初始化状态
     */
    private var init = false

    /**
     * 应用程序名(在文件管理器中的程序文件夹名)
     */
    private var INTERNAL_APP_NAME: String = UNNAMED_APP_DIR

    private var SD_APP_NAME: String = UNNAMED_APP_DIR

    /* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
     *
     * 公开方法
     *
     * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

    /**
     * 初始化
     *
     * @param context 上下文
     */
    fun init(context: Context) {
        var filesDir = context.getExternalFilesDir("")
        val parent: String?
        if (filesDir != null) {
            parent = filesDir.parent
            if (parent != null) {
                SD_APP_NAME = parent
            }
        }
        filesDir = context.filesDir
        INTERNAL_APP_NAME = filesDir.absolutePath
        init = true
    }

    /**
     * 获取本项目文件目录
     *
     * @return 本项目文件目录
     */
    val appDir: File?
        get(){
            if (!init) {
                throw NullPointerException("FileUtil not init")
            }
            if (INTERNAL_APP_NAME.isEmpty()) {
                throw ExceptionInInitializerError("File name invalid")
            }
            val file = File(INTERNAL_APP_NAME)
            val mkdirs: Boolean
            if (!file.exists()) {
                mkdirs = file.mkdirs()
                return if (mkdirs) {
                    file
                } else {
                    null
                }
            }
            return file
        }

    /**
     * 获取本项目在SD卡上的文件目录
     *
     * @return 本项目在SD卡上的文件目录
     */
    val sdCardAppDir: File?
        get() {
            if (!init) {
                throw NullPointerException("FileUtil not init")
            }
            if (SD_APP_NAME.isEmpty()) {
                throw ExceptionInInitializerError("File name invalid")
            }
            val file = File(SD_APP_NAME)
            val mkdirs: Boolean
            if (!file.exists()) {
                mkdirs = file.mkdirs()
                return if (mkdirs) {
                    file
                } else {
                    null
                }
            }
            return file
        }

    /**
     * 获取项目缓存文件目录
     *
     * @return 项目缓存文件目录
     */
    val cacheDir: File?
        get() {
            val file = File(appDir, "cache")
            val mkdirs: Boolean
            if (!file.exists()) {
                mkdirs = file.mkdirs()
                return if (mkdirs) {
                    file
                } else {
                    null
                }
            }
            return file
        }

    /**
     * 获取项目在SD卡上的缓存文件目录
     *
     * @return 项目在SD卡上的缓存文件目录
     */
    val sdCardCacheDir: File?
        get() {
            val file = File(sdCardAppDir, "cache")
            val mkdirs: Boolean
            if (!file.exists()) {
                mkdirs = file.mkdirs()
                return if (mkdirs) {
                    file
                } else {
                    null
                }
            }
            return file
        }

    /**
     * 获取项目的异常日志目录
     *
     * @return 项目的异常日志目录
     */
    val crashDir: File?
        get() {
            val file = File(appDir, "crash")
            val mkdirs: Boolean
            if (!file.exists()) {
                mkdirs = file.mkdirs()
                return if (mkdirs) {
                    file
                } else {
                    null
                }
            }
            return file
        }

    /**
     * 获取项目在SD卡上的异常日志目录
     *
     * @return 项目在SD卡上的异常日志目录
     */
    val sdCardCrashDir: File?
        get() {
            val file = File(sdCardAppDir, "crash")
            val mkdirs: Boolean
            if (!file.exists()) {
                mkdirs = file.mkdirs()
                return if (mkdirs) {
                    file
                } else {
                    null
                }
            }
            return file
        }

    /**
     * 获取apk存储目录
     *
     * @return apk存储目录
     */
    val apkDir: File?
        get() {
            val file = File(appDir, "apk")
            val mkdirs: Boolean
            if (!file.exists()) {
                mkdirs = file.mkdirs()
                return if (mkdirs) {
                    file
                } else {
                    null
                }
            }
            return file
        }

    /**
     * 获取在SD卡上的apk存储目录
     *
     * @return 在SD卡上的apk存储目录
     */
    val sdCardApkDir: File?
        get() {
            val file = File(sdCardAppDir, "apk")
            val mkdirs: Boolean
            if (!file.exists()) {
                mkdirs = file.mkdirs()
                return if (mkdirs) {
                    file
                } else {
                    null
                }
            }
            return file
        }

    /**
     * 获取图片目录
     *
     * @return 图片目录
     */
    val imageDir: File?
        get() {
            val file = File(appDir, "images")
            val mkdirs: Boolean
            if (!file.exists()) {
                mkdirs = file.mkdirs()
                return if (mkdirs) {
                    file
                } else {
                    null
                }
            }
            return file
        }

    /**
     * 获取在SD卡上的图片目录
     *
     * @return 在SD卡上的图片目录
     */
    val sdCardImageDir: File?
        get() {
            val file = File(sdCardAppDir, "images")
            val mkdirs: Boolean
            if (!file.exists()) {
                mkdirs = file.mkdirs()
                return if (mkdirs) {
                    file
                } else {
                    null
                }
            }
            return file
        }

    /**
     * 获取 Log 日志文件保存目录
     *
     * @return Log 日志文件保存目录
     */
    val sdcardLogInfoDir: File?
        get() {
            val file = File(sdCardAppDir, "logInfo")
            val mkdirs: Boolean
            if (!file.exists()) {
                mkdirs = file.mkdirs()
                return if (mkdirs) {
                    file
                } else {
                    null
                }
            }
            return file
        }
    val logInfoDir: File?
        get() {
            val file = File(appDir, "logInfo")
            val mkdirs: Boolean
            if (!file.exists()) {
                mkdirs = file.mkdirs()
                return if (mkdirs) {
                    file
                } else {
                    null
                }
            }
            return file
        }

    /**
     * 获取项目数据目录
     *
     * @return 项目数据目录
     */
    val dataDir: File?
        get() {
            val file = File(appDir, "data")
            val mkdirs: Boolean
            if (!file.exists()) {
                mkdirs = file.mkdirs()
                return if (mkdirs) {
                    file
                } else {
                    null
                }
            }
            return file
        }

    /**
     * 获取项目在SD卡上的数据目录
     *
     * @return 项目在SD卡上的数据目录
     */
    val sdCardDataDir: File?
        get() {
            val file = File(sdCardAppDir, "data")
            val mkdirs: Boolean
            if (!file.exists()) {
                mkdirs = file.mkdirs()
                return if (mkdirs) {
                    file
                } else {
                    null
                }
            }
            return file
        }

    /**
     * 删除文件
     *
     * @param fileName 文件名
     */
    fun delFile(fileName: String): Boolean {
        val file = File(appDir, fileName)
        if (file.isFile) {
            file.delete()
            return true
        }
        file.exists()
        return false
    }

    /**
     * 保存图片
     *
     * @param bm      位图图片
     * @param picName 文件名
     */
    fun saveBitmap(bm: Bitmap, picName: String) {
        Log.e("", "保存图片")
        try {
            val f = File(cacheDir, "$picName.JPEG")
            if (f.exists()) {
                f.delete()
            }
            val out = FileOutputStream(f)
            bm.compress(Bitmap.CompressFormat.JPEG, 90, out)
            out.flush()
            out.close()
            Log.e("", "已经保存")
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    /**
     * 检查文件是否存在
     *
     * @param fileName 文件名
     * @return true表示文件存在
     */
    fun isFileExist(fileName: String): Boolean {
        val file = File(appDir, fileName)
        file.isFile
        return file.exists()
    }

    /**
     * 检查文件是否存在
     *
     * @param path 文件路径
     * @return true表示文件存在
     */
    fun fileIsExists(path: String): Boolean {
        try {
            val file = File(path)
            if (!file.exists()) {
                return false
            }
        } catch (e: Exception) {
            return false
        }
        return true
    }

    /**
     * 删除一个目录
     *
     * @param dir 目录
     */
    fun deleteDirFiles(dir: File?) {
        if (dir == null) {
            return
        }
        if (!dir.exists()) {
            return
        }
        if (!dir.isDirectory) {
            return
        }
        val files = dir.listFiles() ?: return
        for (file in files) {
            if (file.isFile) {
                file.delete() // 删除所有文件
            } else if (file.isDirectory) {
                deleteDirFiles(file) //递规的方式删除文件夹
            }
        }
        dir.delete() // 删除目录本身
    }
}